Quda interface

m-schroeck opened the pull request:
This is an interface for the QUDA library. So far only the "invert" executable supports inversions with Quda. Extension to the determinant monomials for the HMC is straightforward.

See the documentation (tex) for details!

urbach commented:
Hi Mario, could you please merge 'master' into this branch. Thanks!

urbach commented on config.h.in:
```diff
@@ -1,34 +1,135 @@
 /* config.h.in.  Generated from configure.in by autoheader.  */
-#ifndef _CONFIG_H
```
hmm, why was this removed?

m-schroeck replied:
I had to regenerate it with autoheader. Is there a problem? Should we revert it to the previous state?

urbach replied:
ah, okay, never used `autoheader`. We should definitly have the `#ifdef _CONFIG_H`. Otherwise, did you have to add so many new things to `config.h.in`? Do we need all the changes?

urbach commented on doc/main.bbl:
```diff
@@ -0,0 +1,195 @@
+\begin{thebibliography}{10}
```
we don't need main.bbl under git control. It is generated via `bibtex`

m-schroeck replied:
removed.

urbach commented on operator.c:
```diff
@@ -264,12 +278,20 @@ void op_invert(const int op_id, const int index_start, const int write_prop) {
-				      0, optr->even_odd_flag,optr->no_extra_masses, optr->extra_masses, optr->solver_params, optr->id );
-	
+	if(optr->use_qudainverter) {
+#ifdef QUDA
```
wouldn't this switch be better at a lower level? e.g. in `invert_eo.c`?

urbach commented on read_input.l:
```diff
@@ -727,6 +730,17 @@ inline void rmQuotes(char *str){
   }
 }
 
+<WILSONOP,TMOP,DBTMOP,CLOVEROP,DBCLOVEROP>{  
```
how does this fit with the `BeginGPUInit` section in the input file?

m-schroeck replied:
The BeginGPUInit is from Florian Burger, so far I'm not using that. Do you want me to adapt it?

urbach replied:
hmm, what do you think? It offers more possibilities.

I think a combination of the two would be best. Currently this is completely independent, right?

urbach commented:
line indentation is also different compare to the rest. We follow GNU with 4 spaces...

urbach commented:
does this also work with a scalar build, i.e. without an MPI compiler?

urbach commented on quda_interface.c:
```diff
@@ -0,0 +1,786 @@
+// reorder spinor to QUDA format
+void reorder_spinor_toQuda( double* spinor, QudaPrecision precision, int doublet, double* spinor2 )
+{
+	double startTime = MPI_Wtime();
```
use the custom `gettime` please, defined in `gettime.h` / `gettime.c`.

m-schroeck replied:
thanks for adding replacing the timing function etc. 
I removed the tabs in operator.c and quda_interface.c

urbach commented on quda_interface.c:
```diff
@@ -0,0 +1,786 @@
+double *tempSpinor;
+
+// function that maps coordinates in the communication grid to MPI ranks
+int commsMap(const int *coords, void *fdata)
```
maybe this better becomes a generally available function?

m-schroeck replied:
right, could be useful elsewhere (e.g. in the qphix interface). where do you suggest do put it?

urbach commented on quda_interface.c:
```diff
@@ -0,0 +1,786 @@
+
+	int rank;
+
+	MPI_Cart_rank( g_cart_grid, n, &rank );
```
bracket with `#ifdef MPI`

urbach commented:
it seems I can at least compile without MPI compiler. Will it crash?

urbach commented:
please don't use tabs... (even if still sometimes present in the code). E.g. emacs knows `untabbify` as a command...

m-schroeck commented:
Good point with the scalar build! There was in fact an issue (setting device has to be done explicitly in a scalar build). Fixed that. Just tested a scalar QUDA build linked to scalar tmLQCD (all with gcc, g++), compiles and runs fine! Added a note to the doc.

urbach commented:
you mean you have to chose the device ID explicitly? This does not work on our cluster...

kostrzewa commented:
I would like to pull this in now to prevent problems in the next few merges due to conflicts. For now we will simply live with the fact that in serial builds, device 0 is used by default.

kostrzewa commented:
I'm a little bit hesitant because of the "default sloppy precision" in the operator, I don't think we want this in general, do we? I'd rather have the "less safe" version optional.

kostrzewa commented:
Okay, I'll pull it in then, thanks for the clarification. I agree that mixed solvers with reliable updates are the way to go in general.
